docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+Like the limit operator, the unique operator is allowed on log stream without parameters such as `{job="prod/query-frontend"} |= "err" | uniq` will returns the first 1000 unique lines containing the `err` word.
+This is really useful to reduce noise applications might generate by failing constantly.
+
+Using unique operator with metric queries will result into an error.(e.g `rate({job="prod/query-frontend"} | uniq [5m])`).
+
+### Metrics
+
+The first interesting change will be that `count_over_time` and `rate` is now able to count field or log entries. Fields can also be used like labels during metric aggregation.
+
+For example you can now aggregate by level without indexing it using the query below:
+
+```logql
+rate({job="prod/query-frontend"} | logfmt level [1m])
+```
+
+This will automatically add the level fields as part of the series for each series. Again you should be mindful about which fields you select as this will increase the amount of series returned.
+
+Fields can also be used during dimension aggregation like  labels:
+
+```logql
+sum (
+    rate({job="prod/query-frontend"} | logfmt | select path status_code [1m])
+) by (instance, status_code, path)
+
+```
+
+Aggregation dimension will be limited, if fields contains too many unique values we will return an error.
+
+While `count_over_time` and `rate` count entries occurrence per log & field stream, now that we can select fields we will allow some new aggregation functions over those field values.
+
+#### Series & Histogram Operators
+
+To transform fields into series we will introduce two new field stream operators `| series` and `| histogram`. In the future we could introduce other operators like `| counter <field> [inc|dec] by (<label>,<field>)` to sum values if needed.
+
+The series operator `| series <field> by (<label>,<field>)` creates a series from a field stream, the first parameter is the field to use as value, this means each field entry (field/timestamp pair) will create a point (value/timestamp pair). You can then use `by` or `without` to group series metric (again you can include labels or field or both) prior aggregations. You can also omit the grouping clause in which case all extracted fields and labels will define the set of series returned.
+
+Example of a series transformation:
+
+```logql
+{job="app"} | json | series latency by (path, status)
+{job="app"} | json | series latency
+{job="app"} | json | series latency without (instance)
+```
+
+The histogram operator `| histogram <field> [<array_of_float>] by (<label>,<field>)` creates a [Prometheus histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) with the given `<array_of_float>` buckets. It will compute cumulative counters for the observation buckets as well as a counter of all observed values and the total sum of all observations. Series count per query will also be limited, so trying to use too many buckets will result in an error.
```

Comment:
Have you considered `histogram <field> buckets(<array_of_float>) by (<label>,<field>)`? Even if a bit more verbose, I think it would fit better along with `exp(20000, 2, 10)` and `linear(.75, 2, 10)`.

Reply:
I like the consistency of it. I'll make the change.